test-runner-poc/config.js

```diff
@@ -0,0 +1,93 @@
+'use strict';
+const f = require('util').format;
+const url = require('url');
+const qs = require('querystring');
+const core = require('../lib/core');
+class NativeConfiguration {
+  constructor(environment) {
+    this.options = environment || {};
+    this.host = environment.host || 'localhost';
+    this.port = environment.port || 27017;
+    this.db = environment.db || 'integration_tests';
+    this.mongo = environment.mongo;
+    this.setName = environment.setName || 'rs';
+    this.require = this.mongo;
+    this.writeConcern = function() {
+      return { w: 1 };
+    };
+    this.topology = environment.topology || this.defaultTopology;
+    this.environment = environment;
+    if (environment.setName) {
+      this.replicasetName = environment.setName || 'rs';
+    }
+  }
+  usingUnifiedTopology() {
+    return !!process.env.MONGODB_UNIFIED_TOPOLOGY;
+  }
+  newClient(dbOptions, serverOptions) {
+    // support MongoClient contructor form (url, options) for `newClient`
+    if (typeof dbOptions === 'string') {
+      return new this.mongo.MongoClient(
+        dbOptions,
+        this.usingUnifiedTopology()
+          ? Object.assign({ useUnifiedTopology: true }, serverOptions)
+          : serverOptions
+      );
+    }
+    dbOptions = dbOptions || {};
+    serverOptions = Object.assign({}, { haInterval: 100 }, serverOptions);
+    if (this.usingUnifiedTopology()) serverOptions.useUnifiedTopology = true;
+    // Set up the options
+    const keys = Object.keys(this.options);
+    if (keys.indexOf('sslOnNormalPorts') !== -1) serverOptions.ssl = true;
+    // Fall back
+    let dbHost = (serverOptions && serverOptions.host) || 'localhost';
+    const dbPort = (serverOptions && serverOptions.port) || this.options.port || 27017;
+    if (dbHost.indexOf('.sock') !== -1) {
+      dbHost = qs.escape(dbHost);
+    }
+    if (this.options.setName) {
+      Object.assign(dbOptions, { replicaSet: this.options.setName, auto_reconnect: false });
+    }
+    const connectionString = url.format({
+      protocol: 'mongodb',
+      slashes: true,
+      hostname: dbHost,
+      port: dbPort,
+      query: dbOptions,
+      pathname: '/'
+    });
+    return new this.mongo.MongoClient(connectionString, serverOptions);
+  }
+  newTopology(host, port, options) {
+    options = options || {};
+    return this.topology(host, port, options);
+  }
+  url(username, password) {
+    const url = this.options.url || 'mongodb://%slocalhost:27017/' + this.db;
+    // Fall back
+    const auth = username && password ? f('%s:%s@', username, password) : '';
```

Comment:
Can we remove the use of `f` here?

Reply:
The `url` function uses `f`

Reply:
Could we rewrite this so we don't use `f`?

Reply:
rewritten